### ISSUE TITLE
RootProject with relative base throws exception in Windows

### ISSUE THREAD

**Issue by hamjad** [sbt/sbt#105]
To reproduce: given projects A and B in sibling folders a and b, with B an external dependancy of A, having something like 

``` scala
val dep = RootProject( file( "../b" ) )
```

in the build config of A will throw an exception (reproduced below). I think this is because, in case of a relative path, IO.toURI passes the output of File.getPath to the URI constructor. Unfortunately, on Windows,  the string returned by getPath contains backslashes (regardless of how the original path string was specified above) and the URI constructor does not like this. 

Workaround was 

``` scala
val dep = ProjectRef( uri( "../b"), "B" )
```

at which point I lost motivation to find a fix. 

Running sbt 0.10 (not nightly), scala 2.9.0-1, jvm 6u26 (64 bit), Windows 7 (64 bit). The full exception:

Caused by: java.net.URISyntaxException: Illegal character in path at index 2: ..\sbtdemo5
        at java.net.URI$Parser.fail(URI.java:2809)
        at java.net.URI$Parser.checkChars(URI.java:2982)
        at java.net.URI$Parser.parseHierarchical(URI.java:3066)
        at java.net.URI$Parser.parse(URI.java:3024)
        at java.net.URI.<init>(URI.java:578)
        at sbt.IO$.toURI(IO.scala:616)
        at sbt.RootProject$.apply(Reference.scala:31)
        at CCPBuild$.<init>(Build.scala:40)
        at CCPBuild$.<clinit>(Build.scala)
        at java.lang.Class.forName0(Native Method)
        at java.lang.Class.forName(Class.java:247)
        at sbt.ModuleUtilities$.getObject(ModuleUtilities.scala:10)
        at sbt.Load$.loadDefinition(Load.scala:364)
        at sbt.Load$$anonfun$loadDefinitions$1.apply(Load.scala:362)
        at sbt.Load$$anonfun$loadDefinitions$1.apply(Load.scala:362)
        at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:206)
        at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:206)
        at scala.collection.LinearSeqOptimized$class.foreach(LinearSeqOptimized.scala:61)
        at scala.collection.immutable.List.foreach(List.scala:45)
        at scala.collection.TraversableLike$class.map(TraversableLike.scala:206)
        at scala.collection.immutable.List.map(List.scala:45)
        at sbt.Load$.loadDefinitions(Load.scala:362)
        at sbt.Load$.definitions(Load.scala:357)
        at sbt.Load$.loadUnit(Load.scala:327)
        at sbt.Load$$anonfun$5.apply(Load.scala:172)
        at sbt.Load$$anonfun$5.apply(Load.scala:172)
        at sbt.BuildLoader.apply(BuildLoader.scala:17)
        at sbt.Load$.loadAll(Load.scala:224)
        at sbt.Load$.loadURI(Load.scala:181)
        at sbt.Load$.load(Load.scala:177)
        at sbt.Load$.load(Load.scala:175)
        at sbt.Load$.apply(Load.scala:83)
        at sbt.Load$.defaultLoad(Load.scala:41)
        at sbt.BuiltinCommands$$anonfun$loadProjectImpl$2.apply(Main.scala:371)
        at sbt.BuiltinCommands$$anonfun$loadProjectImpl$2.apply(Main.scala:368)
        at sbt.Command$$anonfun$applyEffect$1$$anonfun$apply$2.apply(Command.scala:61)
        at sbt.Command$$anonfun$applyEffect$1$$anonfun$apply$2.apply(Command.scala:61)
        at sbt.Command$$anonfun$applyEffect$2$$anonfun$apply$3.apply(Command.scala:63)
        at sbt.Command$$anonfun$applyEffect$2$$anonfun$apply$3.apply(Command.scala:63)
        at sbt.Command$.process(Command.scala:91)
        at sbt.MainLoop$$anonfun$next$1$$anonfun$apply$1.apply(Main.scala:72)
        at sbt.MainLoop$$anonfun$next$1$$anonfun$apply$1.apply(Main.scala:72)
        at sbt.State$$anon$1.process(State.scala:54)
        at sbt.MainLoop$$anonfun$next$1.apply(Main.scala:72)
        at sbt.MainLoop$$anonfun$next$1.apply(Main.scala:72)
        at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:13)
        at sbt.MainLoop$.next(Main.scala:72)
        at sbt.MainLoop$.run(Main.scala:63)
        at sbt.xMain.run(Main.scala:33)
        at xsbt.boot.Launch$.run(Launch.scala:54)
        at xsbt.boot.Launch$$anonfun$explicit$1.apply(Launch.scala:43)
        at xsbt.boot.Launch$.launch(Launch.scala:68)
        at xsbt.boot.Launch$.apply(Launch.scala:14)
        at xsbt.boot.Boot$.runImpl(Boot.scala:24)
        at xsbt.boot.Boot$.main(Boot.scala:15)
        at xsbt.boot.Boot.main(Boot.scala)

**Comment by benhutchison** [comment]
Although 0.10.1 has a fix included, I dont think this issue is resolved by it. See session below. Perhaps some side-effect of '\' conflicted dual role as path delimiter and string escape character on Windows?

> sbt-version
> [info] 0.10.1
> console-project
> [info] Starting scala interpreter...
> [info]
> import sbt._
> import Process._
> import Keys._
> import currentState._
> import extracted._
> Welcome to Scala version 2.8.1.final (Java HotSpot(TM) 64-Bit Server VM, Java 1.
> 7.0).
> Type in expressions to have them evaluated.
> Type :help for more information.

scala> sbt.IO.toURI(new java.io.File("../tagtree"))
java.net.URISyntaxException: Illegal character in path at index 2: ..\tagtree
        at java.net.URI$Parser.fail(Unknown Source)
        at java.net.URI$Parser.checkChars(Unknown Source)
        at java.net.URI$Parser.parseHierarchical(Unknown Source)
        at java.net.URI$Parser.parse(Unknown Source)
        at java.net.URI.<init>(Unknown Source)
        at sbt.IO$.toURI(IO.scala:620)
        at .<init>(<console>:23)
        at .<clinit>(<console>)
        at RequestResult$.<init>(<console>:9)
        at RequestResult$.<clinit>(<console>)
        at RequestResult$scala_repl_result(<console>)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
        at java.lang.reflect.Method.invoke(Unknown Source)
        at scala.tools.nsc.Int...

**Comment by harrah** [comment]
That is strange.  Here are the relevant methods:

```
def toURI(f: File): URI  =  if(f.isAbsolute) f.toURI else new URI(normalizeName(f.getPath))

def normalizeName(name: String) = {
  val sep = File.separatorChar
  if(sep == '/') name else name.replace(sep, '/')
}
```

You can see that it should be replacing \ with / on windows for relative paths before sending the String to URI.

**Comment by benhutchison** [comment]
Yeah, strange. Ive tested all the pieces above in isolation in the REPL and they work correctly. 

Thus, the only way I can explain it is that Im running a mix of 0.10.0 (sbt.IO) and 0.10.1 (launcher/version) SBT code. Is that plausible?

**Comment by harrah** [comment]
Hmm, thanks.

I think it is more likely that somehow there are both 0.10.0 and 0.10.1 classes on the classpath and the 0.10.0 ones are getting used.  Are you running in a blank project without a ~/.sbt/ directory?

**Comment by benhutchison** [comment]
Further investigation lead me to believe that my io_2.8.1-0.10.1.jar file doesnt contain the 0.10.1 version of the SBT source code.

Here's my method and evidence:
1. Added "-verbose:class" to JVM launch params, and noted the following output:

[Loaded sbt.IO$ from file:/C:/Users/ben/workspace/TagTree/project/boot/scala-2.8.1/org.scala-tools.sbt/sbt/0.10.1/io_2.8.1-0.10.1.jar]

So we are running out of 0.10.1 tagged jars. Also noted no classes were loaded from 0.10.0 tagged jars. 
1. Examined the "sbt.IO$" class bytecode from " io_2.8.1-0.10.1.jar", using this tool [http://andrei.gmxhome.de/bytecode/]

Here is what I find, with no normalizeName() method present:

// Method descriptor #1248 (Ljava/io/File;)Ljava/net/URI;
  // Stack: 3, Locals: 2
  public java.net.URI toURI(java.io.File f);
     0  aload_1 [f]
     1  invokevirtual java.io.File.isAbsolute() : boolean [1273]
     4  ifeq 14
     7  aload_1 [f]
     8  invokevirtual java.io.File.toURI() : java.net.URI [1252]
    11  goto 25
    14  new java.net.URI [202]
    17  dup
    18  aload_1 [f]
    19  invokevirtual java.io.File.getPath() : java.lang.String [1274]
    22  invokespecial java.net.URI(java.lang.String) [207]
    25  areturn
      Line numbers:
        [pc: 0, line: 620]
      Local variable table:
        [pc: 0, pc: 26] local: this index: 0 type: sbt.IO.
        [pc: 0, pc: 26] local: f index: 1 type: java.io.File

**Comment by harrah** [comment]
It isn't there for 0.10.1 for me either, although it is for my locally built 0.10.2-SNAPSHOT.  I'm not sure how that commit didn't get included in the 0.10.1 binary, but since it is in the local binaries, I'll close this.
